src/integrations/prefect-redis/prefect_redis/ordering.py

```diff
@@ -0,0 +1,277 @@
+"""
+Manages the partial causal ordering of events for a particular consumer.  This module
+maintains a buffer of events to be processed, aiming to process them in the order they
+occurred causally.
+"""
+
+import asyncio
+from contextlib import asynccontextmanager
+from datetime import datetime, timedelta, timezone
+from typing import Any, AsyncGenerator, Union
+from uuid import UUID, uuid4
+
+import anyio
+
+from prefect.logging import get_logger
+from prefect.server.events.ordering import (
+    MAX_DEPTH_OF_PRECEDING_EVENT,
+    PRECEDING_EVENT_LOOKBACK,
+    SEEN_EXPIRATION,
+    EventArrivedEarly,
+    MaxDepthExceeded,
+    event_handler,
+)
+from prefect.server.events.ordering import (
+    CausalOrdering as _CausalOrdering,
+)
+from prefect.server.events.schemas.events import Event, ReceivedEvent
+from prefect_redis.client import get_async_redis_client
+
+logger = get_logger(__name__)
+
+
+# How long we'll wait for an in-flight event to be processed for follower handling,
+# which crucially needs to be lower than the stream ack deadline
+IN_FLIGHT_EVENT_TIMEOUT = timedelta(seconds=8)
+
+
+class EventBeingProcessed(Exception):
+    """Indicates that an event is currently being processed and should not be processed
+    until it is finished.  This may happen due to Pub/Sub redelivering a message."""
```

Comment:
nit
```suggestion
    """Indicates that an event is currently being processed and should not be processed
    until it is finished.  This may happen due to Redis Streams redelivering a message."""
```